File: src/main/java/dev/esophose/discordbot/webhook/WebhookUtils.java
package dev.esophose.discordbot.webhook;

import dev.esophose.discordbot.Sparky;
import discord4j.core.object.entity.Webhook;
import discord4j.core.object.entity.channel.MessageChannel;
import discord4j.core.object.entity.channel.TextChannel;
import discord4j.core.object.util.Image;
import discord4j.rest.request.Router;
import discord4j.rest.route.Routes;
import discord4j.rest.service.RestService;
import java.lang.reflect.Method;
import java.util.Objects;
import java.util.function.Consumer;
import reactor.core.publisher.Mono;

public final class WebhookUtils {

    private static Router router;

    static {
        try {
            Method method_getRouter = RestService.class.getDeclaredMethod("getRouter");
            method_getRouter.setAccessible(true);
            router = (Router) method_getRouter.invoke(Sparky.getInstance().getDiscord().getCoreResources().getRestClient().getWebhookService());
        } catch (ReflectiveOperationException e) {
            e.printStackTrace();
        }
    }

    public static Mono<Void> createAndExecuteWebhook(MessageChannel channel, String username, Image avatar, Consumer<? super WebhookExecuteSpec> spec) {
        return ((TextChannel) channel).createWebhook(webhookCreateSpec -> {
            webhookCreateSpec.setName(username);
            webhookCreateSpec.setAvatar(avatar);
        }).flatMap(webhook -> executeWebhook(webhook, spec).then(webhook.delete()));
    }

    public static Mono<Void> createAndExecuteWebhook(Mono<MessageChannel> channelMono, String username, Image avatar, Consumer<? super WebhookExecuteSpec> spec) {
        return channelMono.cast(TextChannel.class).flatMap(channel -> createAndExecuteWebhook(channel, username, avatar, spec));
    }

    public static Mono<Void> executeWebhook(Webhook webhook, Consumer<? super WebhookExecuteSpec> spec) {
        WebhookExecuteSpec mutatedSpec = new WebhookExecuteSpec();
        spec.accept(mutatedSpec);

        MultipartWebhookRequest request = mutatedSpec.asRequest();
        return Routes.WEBHOOK_EXECUTE.newRequest(webhook.getId().asLong(), webhook.getToken())
                .header("content-type", request.getFiles().isEmpty() ? "application/json" : "multipart/form-data")
                .body(Objects.requireNonNull(request.getFiles().isEmpty() ? request.getExecuteRequest() : request))
                .exchange(router)
                .bodyToMono(Void.class);
    }

}

File: src/main/kotlin/dev/esophose/discordbot/webhook/WebhookUtils.kt
package dev.esophose.discordbot.webhook

import dev.esophose.discordbot.Sparky
import discord4j.core.`object`.entity.Webhook
import discord4j.core.`object`.entity.channel.MessageChannel
import discord4j.core.`object`.entity.channel.TextChannel
import discord4j.rest.request.Router
import discord4j.rest.route.Routes
import discord4j.rest.service.RestService
import discord4j.rest.util.Image
import reactor.core.publisher.Mono
import java.util.Objects

object WebhookUtils {

    private var router: Router? = null

    init {
        try {
            val method_getRouter = RestService::class.java.getDeclaredMethod("getRouter")
            method_getRouter.isAccessible = true
            router = method_getRouter.invoke(Sparky.discord.restClient.webhookService) as Router
        } catch (e: ReflectiveOperationException) {
            e.printStackTrace()
        }
    }

    fun createAndExecuteWebhook(channel: MessageChannel, username: String, avatar: Image, spec: (WebhookExecuteSpec) -> Unit): Mono<Void> {
        return (channel as TextChannel).createWebhook { webhookCreateSpec ->
            webhookCreateSpec.setName(username)
            webhookCreateSpec.setAvatar(avatar)
        }.flatMap { webhook -> executeWebhook(webhook, spec).then(webhook.delete()) }
    }

    fun createAndExecuteWebhook(channelMono: Mono<MessageChannel>, username: String, avatar: Image, spec: (WebhookExecuteSpec) -> Unit): Mono<Void> {
        return channelMono.cast(TextChannel::class.java).flatMap { channel -> createAndExecuteWebhook(channel, username, avatar, spec) }
    }

    fun executeWebhook(webhook: Webhook, spec: (WebhookExecuteSpec) -> Unit): Mono<Void> {
        val mutatedSpec = WebhookExecuteSpec()
        spec(mutatedSpec)

        val request = mutatedSpec.asRequest()
        return Routes.WEBHOOK_EXECUTE.newRequest(webhook.id.asLong(), webhook.token)
                .header("content-type", if (request.files.isEmpty()) "application/json" else "multipart/form-data")
                .body(Objects.requireNonNull(if (request.files.isEmpty()) request.executeRequest else request))
                .exchange(router!!)
                .bodyToMono(Void::class.java)
    }

}
